### PR TITLE
[ENH] Prediction intervals refactor: BATS/TBATS; bugfix for #1625; base class updates on `predict_quantiles`

### LINKED ISSUE #1625: [BUG] BATS/TBATS In-sample error
**Describe the bug**
Get reference error when insample predictions are done for BATS and TBATS.

**To Reproduce**
See [code gist here](https://gist.github.com/ngupta23/32856ac7116179fbfe95cea3c1e9ad25)

```python


/usr/local/lib/python3.7/dist-packages/sktime/forecasting/base/_base.py:220: FutureWarning: return_pred_int in predict() will be deprecated;please use predict_interval() instead to generate prediction intervals.
  FutureWarning,

---------------------------------------------------------------------------
UnboundLocalError                         Traceback (most recent call last)
<ipython-input-12-0d389f978f91> in <module>()
      1 import numpy as np
----> 2 forecaster.predict(fh=-np.arange(0, len(y)), return_pred_int=True)

/usr/local/lib/python3.7/dist-packages/sktime/forecasting/base/_base.py in predict(self, fh, X, return_pred_int, alpha, keep_old_return_type)
    227                     X=X_inner,
    228                     return_pred_int=return_pred_int,
--> 229                     alpha=alpha,
    230                 )
    231 

/usr/local/lib/python3.7/dist-packages/sktime/forecasting/base/adapters/_tbats.py in _predict(self, fh, X, return_pred_int, alpha)
    136 
    137         if return_pred_int:
--> 138             return y_pred, pred_int
    139         else:
    140             return y_pred

UnboundLocalError: local variable 'pred_int' referenced before assignment


```

**Expected behavior**
Out of sample works ok. Only insample seems to be an issue.

**Versions**

sktime = 0.8.1

### PR DESCRIPTION
**Pull request by k1m190r**
#### Reference Issues/PRs
 Fixes #1625 

#### What does this implement/fix? Explain your changes.
Fixed #1625. 
Implemented TBATS `_predict_quantiles()`.

##### Re-implemented `_has_predict_quantiles_been_refactored()` due to failing tests.

[Previous implementation](https://github.com/alan-turing-institute/sktime/blob/bce815d11fff62b52882f13cd650c88972b08ab6/sktime/forecasting/base/_base.py#L1250):  
```python
def _has_predict_quantiles_been_refactored(self):
    if "_predict_quantiles" in type(self).__dict__.keys():
        return True
    else:
        return False
```
Relied on the expectation that `type(self)` (last class in inheritance chain) would be class implementing `_predict_quantiles()`, and thus `type(self).__dict__.keys()` would have `"_predict_quantiles"`. 

BUT in case of TBATS `type(self)` is [TBATS class](https://github.com/alan-turing-institute/sktime/blob/bce815d11fff62b52882f13cd650c88972b08ab6/sktime/forecasting/tbats.py#L21) which does not re-implement itself but rather inherits implementation from [_TbatsAdapter](https://github.com/alan-turing-institute/sktime/blob/bce815d11fff62b52882f13cd650c88972b08ab6/sktime/forecasting/base/adapters/_tbats.py#L18). Thus `type(self).__dict__.keys()` does **not** have `"_predict_quantiles"`, and returns _undesired_ `False`.

Suggested implementation:
```python
def _has_predict_quantiles_been_refactored(self):
    """Check if specific forecaster implements _predict_quantiles()."""
    base_pq = BaseForecaster._predict_quantiles
    this_pq = self.__class__._predict_quantiles
    # true if self's _predict_quantiles is new implementation
    return base_pq != this_pq

```
Relies on following expectations: `BaseForecaster._predict_quantiles` [is defined ](https://github.com/alan-turing-institute/sktime/blob/bce815d11fff62b52882f13cd650c88972b08ab6/sktime/forecasting/base/_base.py#L1164)(which it is), `self.__class__._predict_quantiles` or equivalently `type(self)._predict_quantiles` always exist (since `BaseForecaster` is expected to be in the inheritance chain). If any class in the inheritance chain from `BaseForecaster` to `type(self)`/(`self.__class__`) has re-implemented `_predict_quantiles` then `base_pq != this_pq` would be _desired_ `True`.

##### Small changes to  `predict()`.

[In previous implementation of `predict()`](https://github.com/alan-turing-institute/sktime/blob/bce815d11fff62b52882f13cd650c88972b08ab6/sktime/forecasting/base/_base.py#L239). This line has expectation of `alpha` to be iterable/ list. While alpha can be scalar/float.

```python
coverage = [1 - a for a in alpha]
```
Suggestion:
```python
alphas = [alpha] if isinstance(alpha, (float, int)) else alpha
coverage = [1 - a for a in alphas]
```
ensures that float/int becomes a list with that float/int.

Also `else` [clause ](https://github.com/alan-turing-institute/sktime/blob/bce815d11fff62b52882f13cd650c88972b08ab6/sktime/forecasting/base/_base.py#L236)was missing definition for `y_pred` that is expected in return:

Added
```python
y_pred = self._predict(self.fh, X=X_inner)
```

##### Small changes to `_convert_new_to_old_pred_int()` due to failing tests.
This one needs a bit of story :)

[test_predict_pred_interval](https://github.com/alan-turing-institute/sktime/blob/bce815d11fff62b52882f13cd650c88972b08ab6/sktime/forecasting/tests/test_all_forecasters.py#L258) failed thus:

[this line](https://github.com/alan-turing-institute/sktime/blob/bce815d11fff62b52882f13cd650c88972b08ab6/sktime/forecasting/tests/test_all_forecasters.py#L282)
```python
y_pred, pred_ints = f.predict(return_pred_int=True, alpha=alpha)
```
where `alpha=0.05`

in [`predict()`](https://github.com/alan-turing-institute/sktime/blob/bce815d11fff62b52882f13cd650c88972b08ab6/sktime/forecasting/base/_base.py#L240)
```python
pred_int = self.predict_interval(fh=fh, X=X_inner, coverage=coverage)
```
where `coverage=[0.95]`

`predict_interval()` calls [`_predict_interval()`](https://github.com/alan-turing-institute/sktime/blob/bce815d11fff62b52882f13cd650c88972b08ab6/sktime/forecasting/base/_base.py#L1157) 

which has these lines

```python
alphas = []
for c in coverage:
    alphas.extend([(1 - c) / 2, 0.5 + (c / 2)])
alphas.sort()
```
after these lines `alphas = [0.025000000000000022, 0.975]`, which become columns for the multiindex returned by [`_predict_quantiles()`](https://github.com/alan-turing-institute/sktime/blob/bce815d11fff62b52882f13cd650c88972b08ab6/sktime/forecasting/base/_base.py#L1161)

**Notice first quantile `0.025000000000000022`!** since `(1 - 0.95) = 0.050000000000000044` not so fun numerical instability :(.

once returned back to `predict()` we are executing [these lines ](https://github.com/alan-turing-institute/sktime/blob/bce815d11fff62b52882f13cd650c88972b08ab6/sktime/forecasting/base/_base.py#L243) to convert from new to old:
```python
if keep_old_return_type:
    pred_int = _convert_new_to_old_pred_int(pred_int, alpha)
```

Now actual **problem**:

[Previous implementation of _convert_new_to_old_pred_int](https://github.com/alan-turing-institute/sktime/blob/bce815d11fff62b52882f13cd650c88972b08ab6/sktime/forecasting/base/_base.py#L1308):

```python
def _convert_new_to_old_pred_int(pred_int_new, alpha):
    name = pred_int_new.columns.get_level_values(0).unique()[0]
    alpha = check_alpha(alpha)
    pred_int_old_format = [
        pd.DataFrame(
            {
                "lower": pred_int_new[name, a / 2],
                "upper": pred_int_new[name, 1 - (a / 2)],
            }
        )
        for a in alpha
    ]

    # for a single alpha, return single pd.DataFrame
    if len(alpha) == 1:
        return pred_int_old_format[0]

    # otherwise return list of pd.DataFrames
    return pred_int_old_format
```

takes original `alpha=0.05` and attempts to construct expected column index name `[name, a/2]` here again `alpha` is expected to be a iterable/list but is float (so fixed).

BUT the main problem is **`a/2=0.25` exactly `0.025` and thus fails to get actual column `0.025000000000000022`**

Suggested implementation, simply replicates the computation path to arrive to exactly the same values of quantiles.

```python
def _convert_new_to_old_pred_int(pred_int_new, alpha):
    name = pred_int_new.columns.get_level_values(0).unique()[0]
    alpha = check_alpha(alpha)

    alphas = [alpha] if isinstance(alpha, (float, int)) else alpha
    coverage = [1 - a for a in alphas]

    alpha_quantiles = []
    for c in coverage:
        aq_pair = ((1 - c) / 2, 0.5 + (c / 2))
        alpha_quantiles.append(aq_pair)
    alpha_quantiles.sort()

    pred_int_old_format = [
        pd.DataFrame(
            {
                "lower": pred_int_new[(name, al)],
                "upper": pred_int_new[(name, au)],
            }
        )
        for al, au in alpha_quantiles
    ]

    # for a single alpha, return single pd.DataFrame
    if len(alphas) == 1:
        return pred_int_old_format[0]

    # otherwise return list of pd.DataFrames
    return pred_int_old_format
```

#### Does your contribution introduce a new dependency? If yes, which one?
No

#### What should a reviewer concentrate their feedback on?
- general code comments (my first commit :))

#### PR checklist

TODO
##### For all contributions
- [ ] Will add myself to the [list of contributors](https://github.com/alan-turing-institute/sktime/blob/main/.all-contributorsrc).

### CODE REVIEW

File: sktime/forecasting/base/adapters/_tbats.py [hunk 1]
```diff
@@ -89,21 +91,23 @@ def _fit(self, y, X=None, fh=None):
...
         """Forecast time series at future horizon.
 
         Parameters
         ----------
         fh : int, list, np.array or ForecastingHorizon
             Forecasting horizon
-        X : pd.DataFrame, optional (default=None)
-            Exogenous time series
+        X : (default=None)
+            NOT USED BY TBATS
```
Comment (fkiraly-shell) [inline sktime/forecasting/base/adapters/_tbats.py:106]: maybe say sth like "argument ignored but present for interface compatibility"?

File: sktime/forecasting/base/adapters/_tbats.py [hunk 2]
```diff
@@ -89,21 +91,23 @@ def _fit(self, y, X=None, fh=None):
...
         return_pred_int : bool, optional (default=False)
             If True, returns prediction intervals for given alpha values.
-        alpha : float or list, optional (default=0.95)
+        alpha : float, optional (default=0.05)
+            Interpreted as "Confidence Interval" = 1 - alpha
```
Comment (fkiraly-shell) [inline sktime/forecasting/base/adapters/_tbats.py:110]: unclear what you mean. As coverage = 1-alpha, or upper/lower quantiles (mirrored)?

File: sktime/forecasting/base/adapters/_tbats.py [hunk 3]
```diff
@@ -188,3 +333,8 @@ def _get_pred_int(self, lower, upper):
         pred_int.index = fh_out.to_absolute(self.cutoff)
         pred_int = pred_int.drop(columns=["idx"])
         return pred_int
+
+
+def nans(length):
+    """Return l NaNs."""
```
Comment (fkiraly-shell) [inline sktime/forecasting/base/adapters/_tbats.py:339]: perhaps don't use the letter `l`, easily confused with the number `1`.

**Review by fkiraly-shell — DISMISSED** [review]
Looks good to me, comments are all docstring related, no major blockers.
Let me know once you have gone through them and/or decided that there is no action needed.

**Comment by fkiraly-shell** [comment]
@kejsitake, @SveaMeyer13, this modifies `_has_predict_quantiles_been_refactored`, and `_convert_new_to_old_pred_int`, if you want to have a look.

@k1m190r, would you kindly be able to quickly explain what you changed and why this needed change? I know we've looked into this to some extent in pair programming, but I think it would be good for @kejsitake and @SveaMeyer13 to understand what's being done to functions they wrote and worked on.

**Comment by k1m190r** [comment]
> 

Edited original PR with details. Hope it helps :)

**Comment by kejsitake** [comment]
Thank you for the explanation. I also noticed the issue with the float values, I solved it by casting the values to float, but this looks cleaner.

In my PR refactoring `predict_quantiles` in theta https://github.com/alan-turing-institute/sktime/pull/1579, I am sorting the alphas, just because I thought it would the output would more readable. But, I can back and change it and was wondering, for the sake of consistency, what does everyone think about this?

**Comment by fkiraly** [comment]
> for the sake of consistency, what does everyone think about this?

I vaguely remember we discussed this and made a decision, although I don't remember anymore what the decision was or where we noted it down. Was it "we sort it"?

**Comment by fkiraly** [comment]
@k1m190r, I've merged @kejsitake's PR https://github.com/alan-turing-institute/sktime/pull/1579 and tried to resolve merge inconsistencies.

Can you kindly check whether everything is still working as intended?

@kejsitake, could you also kindly review whether things make sense?
Some things that are changed from your PR, @kejsitake, to the state in @k1m190r's:

* I removed the renaming of `"Quantiles"` to `"Intervals"` in `predict_intervals` - the numbers are still quantiles, and intervals caption is inaccurate
* The helper function `_convert_new_to_old_pred_int` is now inside the base class
* there was a stray argument `alpha` in `_predict` that you added, I removed that again
* I changed `coverage` from 0.95 to the "right" default 0.90 (this results in the default of 0.05, 0.95 quantiles)

**Comment by fkiraly-shell** [comment]
@k1m190r is off sick, so I'll take this on to get it over the last metres before the finish line.

**Comment by fkiraly** [comment]
(resolved conflicts - merge main into this, then override `BaseForecaster` with main version; resolve typos)

**Review by fkiraly — APPROVED** [review]
Works after resolving simple merge conflicts, all good.
Thanks, @k1m190r.

File: sktime/forecasting/base/adapters/_tbats.py (hunk 1 above)
Comment (fkiraly) [reply sktime/forecasting/base/adapters/_tbats.py:106]: fixed this myself

File: sktime/forecasting/base/adapters/_tbats.py (hunk 2 above)
Comment (fkiraly) [reply sktime/forecasting/base/adapters/_tbats.py:110]: good enough for now

**Pull request merged**